### ISSUE TITLE
clarify and improve availability of older installer versions

### ISSUE THREAD

**Issue by indirectlylit** [learningequality/kolibri-docs#48]
Migrated from https://github.com/learningequality/kolibri/issues/4208

-----

Current state:

* In http://kolibri.readthedocs.io/, we intentionally allow people to view docs for older versions
* If someone intentionally navigates to an older version and tries to download an installer, they are redirected to the most recent version
  * On one hand, this is good because it encourages users to upgrade
  * On the other hand, this is bad because it's misleading, since the user may be intentionally trying to download an older version (e.g. to see if an issue they're having is a regression, or to re-install in a broader implementation that is still using a legacy version)

Needs:

1. Users should be encouraged to upgrade to the latest version whenever they are viewing docs or downloads for older versions, in case they got there unintentionally
2. Users should have an officially-supported and discoverable way to download any previous release, if that's what they really want
3. Downloads should have appropriate analytics tracking (e.g. we get better information from website redirects than through pantry nginx logs)

I don't think this is an urgent problem to solve, but in the long run I think we'll likely want to make a combination of changes to pantry, the website, and the docs to satisfy the three needs above.

**Comment by benjaoming** [comment]
I agree. It cannot be a question of *if* users can land on out-dated documentation because of search engines or links from other sites.. it must be a question of *how many*.

Suggested fix:

1. Do not refer directly to the latest installer. Instead, refer to the download page. New installer types appear.
2. Display the Kolibri version on all the installation instruction pages.

All fixes would benefit from being back-ported to older release branches.

A small note: If the user is displayed the version of Kolibri which the docs were written for, that may still not lead them to conclude "this is out-dated" since she may not be aware of the latest version. However, always referring people to the main download page should...

**Comment by benjaoming** [comment]
Suggesting that we use what RTD offers:

![image](https://user-images.githubusercontent.com/374612/47256331-b3a98800-d47e-11e8-9d94-fd1fc4c31d52.png)

The requirement is that we start tagging the documentation using git tags. ReadTheDocs will automatically add the notification for people browsing old tags.

The disadvantage is that Kolibri is called a "library" in the notification, but we could probably live with this and just try to get that fixed upstream, suggesting to call it something more generic like "software" instead of "library".

**Comment by benjaoming** [comment]
@radinamatic @indirectlylit any comments here?

You can refer to the Wagtail docs for a look at how this ends up... https://docs.wagtail.io/en/latest/

**Comment by radinamatic** [comment]
> start tagging the documentation using git tags

Sounds like a good idea to me, but I don't like the idea of introducing the concept of `library`... We already have the potential uncertainty among the users created by the terms `app` vs. `server` when it comes to Kolibri, and putting the "library" in the mix may make the matters worse.  Specially since "library" is often used in describing content on Studio... :confused: 

I would prefer we put the request for change upstream, and once it is fixed, we start using the feature. For now I [changed the links to point to the main download page instead of the Pantry, and put the explicit reference that it is the link for version 0.11](https://github.com/learningequality/kolibri-docs/pull/54/files#diff-dfe4fde5e9b9c1b47ba261d7d59e83dfR16).

**Comment by benjaoming** [comment]
> Sounds like a good idea to me, but I don't like the idea of introducing the concept of "library"

I think it's okay to compromise because it's an upstream issue (that I will work to have resolved). And the box will appear to guide people in a situation where "no box" is worse than "a box that says 'library'" :)

Radina, I will try to get the upstream change through, and we can decide based upon how long that takes...

**Comment by indirectlylit** [comment]
Regarding possible text for their box, they could change it to reference the docs rather than the software. For example:

> These docs are for an older version. The newest version is [v1.2.3]()

**Comment by benjaoming** [comment]
@indirectlylit yup, brilliant, and I already made the text say that like at the same time as your comment. I guess the intergalactic mental unification waves are playing with us :)

**Comment by benjaoming** [comment]
Here's btw a nice few words on versioning and RTD: https://docs.readthedocs.io/en/latest/versions.html#how-we-envision-versions-working

**Comment by benjaoming** [comment]
Upstream PR: https://github.com/rtfd/readthedocs.org/pull/4842

**Comment by benjaoming** [comment]
The upstream PR is merged, I'll check back periodically to see if it's been deployed...

**Comment by indirectlylit** [comment]
great!

**Comment by indirectlylit** [comment]
https://github.com/rtfd/readthedocs.org/pull/4842 has been merged. 

How do we enable this warning?

**Comment by radinamatic** [comment]
Relevant setting is properly set, so not sure why the banner is not showing... @benjaoming?

![Selection_105](https://user-images.githubusercontent.com/1457929/59660437-ceb4cf00-91a8-11e9-8bd9-e52a4ca11add.png)

**Comment by benjaoming** [comment]
The PR is merged and works in other projects:

![image](https://user-images.githubusercontent.com/374612/59721409-9ac9c000-9221-11e9-824b-9dc2f71f720b.png)

**Comment by radinamatic** [comment]
Then we must be missing something else, will investigate! :+1:

**Comment by benjaoming** [comment]
I've tried rebuilding the 0.11.x docs and nothing shows up. Also tried rebuilding 'latest'.

Something confuses me: I cannot find any settings where we configure which tags and branches are available?

**Comment by benjaoming** [comment]
I simply cannot enable or disable branches nor git tags anymore.. something is broken.

**Comment by indirectlylit** [comment]
do we need to install this extension? https://github.com/humitos/sphinx-version-warning

**Comment by benjaoming** [comment]
@indirectlylit that's an extension which was created to address some short-comings of the universal, built-in RTD solution. The latter should be fine for us, and we won't have update all our branches to get the banner shown.. just gotta figure out why it's not showing :/

**Comment by benjaoming** [comment]
So, the version warning is injected via the footer.

https://github.com/rtfd/readthedocs.org/blob/b67fbf7c9bae4cdf38d05bc93c251fd84352df4e/readthedocs/core/static-src/core/js/doc-embed/footer.js#L74-L79

Looking at the JSON data submitted: https://readthedocs.org/api/v2/footer_html/?callback=jQuery32107480097834923729_1561224561179&project=kolibri&version=release-v0.11.x&page=install&theme=sphinx_rtd_theme&format=jsonp&docroot=/docs/&source_suffix=.rst

...in the response, we have:

```
"show_version_warning":true
```

Thus, `versionCompare.init(data.version_compare)` should be called. Looking there, I find: https://github.com/rtfd/readthedocs.org/blob/master/readthedocs/core/static-src/core/js/doc-embed/version-compare.js#L9

And that's the issue. Because we also have `"is_highest":true` in the JSON response data for the `release-v0.11.x` branch :/

So the question is: Why does Read The Docs API think that `release-v0.11.x` should qualify as the "highest" version?

@radinamatic do you see any settings for selecting branches or tags?

**Comment by benjaoming** [comment]
Okay, I traced where `is_highest` is defined: https://github.com/rtfd/readthedocs.org/blob/master/readthedocs/api/v2/views/footer_views.py#L57

It seems like we could have issues in relation to our branch or tag names. But I can't quite grasp it. The RTD documentation doesn't give any hints about branch and version names: https://docs.readthedocs.io/en/latest/versions.html#how-we-envision-versions-working

**Comment by radinamatic** [comment]
@benjaoming My interpretation of what I can see here is that the version warning banner is activated, and that it should display on any branch that is **not** ``latest``, that is **all but** ``release-v0.12.x``.

![Selection_120](https://user-images.githubusercontent.com/1457929/59970760-7eee5300-956e-11e9-8184-33ca19c4ad5d.png)

I see nothing in the settings that would point to ``release-v0.11.x`` branch being the "highest" :confused:

**Comment by benjaoming** [comment]
@radinamatic but that's what the data says that the 0.11.x branch is using:

Looking at the JSON data submitted: https://readthedocs.org/api/v2/footer_html/?callback=jQuery32107480097834923729_1561224561179&project=kolibri&version=release-v0.11.x&page=install&theme=sphinx_rtd_theme&format=jsonp&docroot=/docs/&source_suffix=.rst

**Comment by indirectlylit** [comment]
`"is_highest":true` is in the response for all branches, not just v0.11.x.

**Comment by benjaoming** [comment]
@indirectlylit yes, however take a look here:

https://docs.wagtail.io/en/v2.4/

And here:

https://readthedocs.org/api/v2/footer_html/?callback=jQuery321015518982107179713_1561402364929&project=wagtail&version=v2.4&page=index&theme=sphinx_rtd_theme&format=jsonp&docroot=/docs/&source_suffix=.rst&_=1561402364930

You can see that RTD has a different opinion about a tag named "v2.4", which makes me wonder. Then I wanted to experiment with out tags and branches.. but I can't find them listed on RTD anymore and was wondering if @radinamatic sees them when logged in?

**Comment by radinamatic** [comment]
@benjaoming I believe tags can be added here: 

https://readthedocs.org/dashboard/kolibri/version/release-v0.12.x/

**Comment by radinamatic** [comment]
![Selection_122](https://user-images.githubusercontent.com/1457929/60044945-ccc89f80-96c3-11e9-88d0-8b6b41a12d27.png)

**Comment by benjaoming** [comment]
@radinamatic I misunderstood the list, it's of course because we don't keep release tags in `kolibri-docs`. I'll add one in order to make a test.

On this page, it should appear as an inactive version.

https://readthedocs.org/projects/kolibri/versions/

**Comment by benjaoming** [comment]
@radinamatic @indirectlylit re-read my old and long comments, which were very useful to pick up on a longer line of inconclusive investigation :)

We can have warnings embedded if we start tagging documentation releases. RTD expects that tags follow software releases in a semantic way and thus prints the name of a branch or a tag:

![image](https://user-images.githubusercontent.com/374612/64027010-2a53ef00-cb40-11e9-882b-8a6a638a4075.png)

However, the concrete version of the software is still maintained in `conf.py` and used "somewhat" correctly here (I said "somewhat" because we don't maintain it rigorously):

![image](https://user-images.githubusercontent.com/374612/64027096-52435280-cb40-11e9-8956-f510a098bbae.png)

Our latest docs say we are running 0.12.7, which is outdated since 0.12.8 was released and we didn't manually bump the version in our `conf.py`.

The repo `kolibri` is tagged and released according to its software releases. The documentation is maintained separately and may be improved after a release. We should DEFINITELY keep this loose dependency between the two.

So how do we:

* make sure that the documentation repo uses git tags in a way that RTD understands?
* track the upstream latest series version number? ("What is the latest 0.12.x"?)
* display links to older installers correctly *or* a GIANT warning that the docs are for version 0.11.x but you'll get the installer of another series?

Since `conf.py` is actually a Python script, we could make it work in the following way:

* Having the correct version in `conf.py`: To specify the minor series, i.e. `0.12`, and to add this to `requirements.txt` as `kolibri>=0.12,<0.13`. In this way, kolibri will be installed and can be found using `kolibri.__version__`. If we are on `develop` with an unreleased kolibri, we can specify this in `conf.py` such that it doesn't attempt to read the version from the PyPi installed dependency.

* Tracking the version with git tags: I would propose that we keep the `release-v0.12.x` branches. This way, we can maintain docs for each of our releases. But we can also add tags to the release branches that RTD can understand and sort. For instance using consecutive numbering such as `v0.12.docs-1` OR by overwriting the a tag each time we want to release a new version - so that the tag for `release-v0.12.x` is always `v0.12.x` but is overwritten at each release (deleted and re-created on Github). Old versions of 0.12.x docs are useless anyways, so old tags might as well be stashed, as to not clutter the RTD settings panels (which will list all known tags).

* In `conf.py`, create a new Sphinx shortcut for download links to the latest installers and possibly a new set of redirects, e.g. `learningequality.org/r/windows-installer-0.12.x`

If you have suggestions or are okay with this, please LMK, I can have a look, and probably go through all the old branches without too much effort.

**Comment by indirectlylit** [comment]
Some quick notes:

1. As you said, it's important for us to be able to continue updating online docs after Kolibri releases
2. On the other hand, eventually we'll eventually begin bundling and distributing the docs with the Kolibri application. For that reason, the Kolibri application will need to be able to pin itself to a specific tagged version of the docs
4. I don't want to start deleting and re-creating tags with the same name, because it would make (2) more confusing.
3. For some version `A.B.C` of Kolibri, the documentation _must_ track `A` and `B`. The benefits of also tracking `C` are a bit smaller.


I like the idea of having a docs tagging convention something like `v0.12-rev1`, `v0.12-rev2`, etc...

>  RTD expects that tags follow software releases in a semantic way and thus prints the name of a branch or a tag

Do you mean that RTD expects a valid Semantic Versioning string as a tag? Or just that it's a "meaningful" tag?

**Comment by benjaoming** [comment]
> Do you mean that RTD expects a valid Semantic Versioning string as a tag? Or just that it's a "meaningful" tag?

Yes, I think that we can rely on RTD creating a properly comparable version string from a tag called `v1.2.3-abc123`. This was where they did the comparison: https://github.com/readthedocs/readthedocs.org/blob/98a7ff1758d829323e5ef9949f57401f7103ec4e/readthedocs/api/v2/views/footer_views.py#L57

I like your suggestion of dropping the patch version numbering from the documentation. We can offload responsibility of fetching the 0.11 installer (linked in the documentation) to for instance a URL called `https://learningequality.org/r/kolibri-deb-latest-0.11.x` and being maintained elsewhere. 

This may have implications on statistics, but any solution we want to implement in the documentation for pointing to specific installer versions would eventually have to NOT point to `https://learningequality.org/r/kolibri-<xxx>-latest`.

If we need the patch version number introduced, we can introduce automation in `conf.py` (for instance to lookup in PyPi), but since it's not necessary, I would propose that the fix was what was outlined in https://github.com/learningequality/kolibri-docs/issues/48#issuecomment-526637503, except for the part of PyPi lookups. And that we don't create a myariade of old tags, but try to stick to a minimal set, that can present users to an overview.

A small hopeful reflection: `0.12-rev3` can make readers and community aware that the documentation is alive and can be improved underway :)

**Comment by benjaoming** [comment]
@radinamatic just wondering if you followed this and if it sounds okay?

**Comment by indirectlylit** [comment]
Bumping this issue because I [noticed someone in the forums](https://community.learningequality.org/t/kolibri-failed-to-start-additional-information-can-be-found/1394/5) linking to [docs from 0.7.x](https://kolibri.readthedocs.io/en/release-v0.7.x/manage.html).

@benjaoming perhaps this is something you can take the lead on?

**Comment by benjaoming** [comment]
@indirectlylit - sure! I also know that Discourse has certain utility functions for cleaning up links and texts in posts, so this might also be an effective way of mitigating this issue.

**Comment by indirectlylit** [comment]
I think the more fundamental problem is actually with how search engines handle versioning, and I'm a bit worried that in fact our banner will not be sufficient – though it's an excellent and necessary start.

The only reason people are posting old links is because that's what they're finding in google. For example, I just guessed at the search that user might have made, and indeed google returns old docs as the top result:

> ![image](https://user-images.githubusercontent.com/2367265/73615045-b370c580-45b9-11ea-9264-768d4efee6db.png)

I encounter this issue regularly when e.g. searching documentation about major Python libraries that have a long history of releases – I imagine you might have had similar experiences?

Perhaps as part of this work we should add a custom `robots.txt` file to prevent old versions from being indexed? see e.g.:

https://docs.readthedocs.io/en/latest/faq.html#how-can-i-avoid-search-results-having-a-deprecated-version-of-my-docs

**Comment by benjaoming** [comment]
Excellent - the docs for readthedocs are really good, and it doesn't seem complicated to develop a recipe for a robots.txt exclusion of old versions. Or perhaps all versions in general, such that only the `/latest` hierarchy is indexed.

> I encounter this issue regularly when e.g. searching documentation about major Python libraries that have a long history of releases – I imagine you might have had similar experiences?

We are one of the more progressive teams, so I'm not surprised to encounter issues first in our own domain :) But I take that the fact that RTD have this documentation section hints that others confirm that it's great to exclude version hierarchies, as their example shows

![image](https://user-images.githubusercontent.com/374612/73615410-f194d600-4607-11ea-9063-7b53ea27c145.png)

**Comment by CloudArtAdmins** [comment]
Did realize there was something open for this.
Made a pull request for debian users.
#160

**Comment by benjaoming** [comment]
@radinamatic a quick note -- since we spoke about this -- the issue seems to be the same: https://www.google.com/search?hl=en&q=kolibri%20docs%20change%20location

The robots.txt is in place as it should:
https://kolibri.readthedocs.io/robots.txt

Since 29 days have passed, I will try to add a "Disallow" pattern...
